## Module 1: Exercises 4: NumPy Basics

---
### Exercise 4.1: Your First NumPy Array
---

You're now going to dive into the world of basketball. 

Along the way, you'll get comfortable with the basics of `numpy`, a powerful package to do data science.

A list `basketball` has already been defined in the Python script, representing the *height* of some basketball players in *centimeters*. Can you add some code to create a `numpy` array from it?

#### Instructions 4.1

*  Import the `numpy` package as `np`, so that you can refer to `numpy` with `np`.

*  Use `np.array()` to create a numpy array from `basketball`. Name this array `np_basketball`.

*  Print out the type of `np_basketball` to check that you got it right.

In [3]:
## code.py 4.1

# Create list basketball
basketball = [180, 215, 210, 210, 188, 176, 209, 200]

# Import the numpy package as np
import numpy as np


# Create a numpy array from basketball: np_basketball
np_basketball = np.array(basketball)

# Print out type of np_basketball
print(type(np_basketball))

<class 'numpy.ndarray'>


---
### Exercise 4.2: Basketball players' height
---

You are a huge basketball fan. 

You decide to call the PBA and ask around for some more statistics on the height of the main players. They pass along data on more than a five hundred players, which is stored as a regular Python list: `height_in`. The height is expressed in *inches*. Can you make a numpy array out of it and convert the units to *meters*?

`height_in` is already available and the `numpy` package is loaded, so you can start straight away (https://en.wikipedia.org/wiki/List_of_Philippine_Basketball_Association_players_(U%E2%80%93Z).



In [34]:
# height in inches of pba players since 1976
height_in = [70, 71, 73, 73, 67, 75, 75, 76, 73, 74, 77, 73, 73, 72, 68, 72, 76,
       67, 78, 77, 76, 73, 77, 73, 73, 80, 71, 70, 76, 79, 73, 77, 76, 80,
       72, 76, 68, 74, 79, 71, 77, 71, 72, 79, 70, 82, 73, 73, 75, 75, 69,
       76, 74, 70, 76, 77, 77, 77, 68, 70, 75, 71, 71, 75, 79, 70, 79, 73,
       76, 78, 76, 70, 72, 72, 76, 73, 81, 76, 77, 73, 79, 81, 71, 72, 78,
       75, 79, 82, 78, 76, 74, 76, 84, 69, 72, 77, 68, 71, 72, 71, 69, 68,
       73, 78, 75, 67, 69, 76, 77, 68, 79, 78, 76, 70, 75, 74, 72, 76, 76,
       67, 76, 71, 71, 70, 74, 73, 67, 77, 71, 72, 73, 72, 77, 72, 70, 78,
       79, 76, 74, 75, 78, 76, 80, 71, 74, 76, 77, 72, 72, 80, 74, 75, 75,
       72, 74, 68, 76, 69, 76, 77, 74, 74, 76, 75, 76, 78, 77, 77, 75, 74,
       71, 77, 71, 76, 75, 76, 75, 74, 75, 77, 75, 72, 79, 78, 71, 75, 74,
       77, 69, 72, 72, 76, 78, 77, 79, 76, 80, 78, 81, 74, 71, 73, 66, 78,
       76, 75, 72, 78, 79, 75, 74, 77, 73, 72, 75, 74, 79, 77, 76, 74, 77,
       75, 78, 78, 73, 76, 77, 75, 79, 78, 76, 74, 84, 80, 70, 76, 76, 75,
       77, 75, 73, 77, 77, 73, 74, 74, 74, 68, 75, 77, 81, 76, 78, 69, 75,
       71, 69, 71, 77, 68, 78, 73, 76, 76, 75, 79, 73, 71, 73, 71, 68, 68,
       74, 76, 76, 76, 76, 78, 76, 77, 78, 69, 78, 73, 69, 75, 74, 72, 70,
       77, 71, 72, 75, 76, 77, 76, 74, 77, 79, 81, 73, 72, 78, 74, 75, 76,
       76, 76, 75, 75, 72, 75, 76, 77, 69, 69, 77, 77, 72, 73, 78, 72, 75,
       69, 76, 77, 72, 74, 70, 81, 72, 76, 73, 75, 73, 74, 72, 77, 74, 79,
       73, 72, 73, 71, 73, 77, 76, 74, 77, 81, 71, 78, 74, 73, 71, 74, 70,
       74, 68, 68, 77, 79, 79, 75, 69, 74, 79, 68, 76, 74, 77, 68, 75, 73,
       71, 74, 74, 76, 71, 77, 76, 74, 74, 78, 76, 73, 76, 75, 76, 73, 74,
       72, 76, 74, 70, 72, 75, 78, 73, 78, 80, 82, 80, 79, 76, 70, 73, 75,
       75, 78, 78, 79, 77, 76, 71, 74, 72, 77, 75, 70, 70, 78, 73, 73, 76,
       72, 65, 69, 76, 71, 76, 76, 75, 74, 76, 83, 75, 75, 73, 75, 76, 74,
       80, 69, 70, 74, 81, 73, 76, 73, 77, 70, 77, 74, 79, 78, 72, 71, 78,
       76, 70, 68, 74, 75, 78, 75, 70, 74, 72, 72, 73, 76, 76, 78, 73, 67,
       70, 74, 74, 71, 73, 74, 76, 78, 72, 78, 79, 81, 70, 76, 77, 71, 74,
       73, 76, 77, 76, 79, 79, 72, 72, 73, 75, 76, 68, 75, 77, 72, 78, 68,
       78, 73, 66, 78, 71, 71]

In [36]:
print(type(height_in))
print(len(height_in))

<class 'list'>
516


#### Instructions 4.2

*  Create a `numpy` array from `height_in`. Name this new array `np_height_in`.

*  Print `np_height_in`.

*  Multiply `np_height_in` with `0.0254` to convert all height measurements from `inches` to `meters`. Store the new values in a new array, `np_height_m`.

*  Print out `np_height_m` and check if the output makes sense.


In [37]:
## code.py 4.2

# Import numpy
import numpy as np

# Create a numpy array from height_in: np_height_in 
np_height_in = np.array(height_in)


# Print out np_height_in
print(np_height_in)


# Convert np_height_in to m: np_height_m
np_height_m = 0.0254 * np_height_in


# Print np_height_m
print(np_height_m)


[70 71 73 73 67 75 75 76 73 74 77 73 73 72 68 72 76 67 78 77 76 73 77 73
 73 80 71 70 76 79 73 77 76 80 72 76 68 74 79 71 77 71 72 79 70 82 73 73
 75 75 69 76 74 70 76 77 77 77 68 70 75 71 71 75 79 70 79 73 76 78 76 70
 72 72 76 73 81 76 77 73 79 81 71 72 78 75 79 82 78 76 74 76 84 69 72 77
 68 71 72 71 69 68 73 78 75 67 69 76 77 68 79 78 76 70 75 74 72 76 76 67
 76 71 71 70 74 73 67 77 71 72 73 72 77 72 70 78 79 76 74 75 78 76 80 71
 74 76 77 72 72 80 74 75 75 72 74 68 76 69 76 77 74 74 76 75 76 78 77 77
 75 74 71 77 71 76 75 76 75 74 75 77 75 72 79 78 71 75 74 77 69 72 72 76
 78 77 79 76 80 78 81 74 71 73 66 78 76 75 72 78 79 75 74 77 73 72 75 74
 79 77 76 74 77 75 78 78 73 76 77 75 79 78 76 74 84 80 70 76 76 75 77 75
 73 77 77 73 74 74 74 68 75 77 81 76 78 69 75 71 69 71 77 68 78 73 76 76
 75 79 73 71 73 71 68 68 74 76 76 76 76 78 76 77 78 69 78 73 69 75 74 72
 70 77 71 72 75 76 77 76 74 77 79 81 73 72 78 74 75 76 76 76 75 75 72 75
 76 77 69 69 77 77 72 73 78 72 75 69 76 77 72 74 70

---
## Exercise 4.3: NumPy Side Effects
---

`numpy` is great for doing vector arithmetic.

If you compare its functionality with regular Python lists, however, some things have changed.

*  First of all, numpy arrays **cannot** contain elements with different types. 
  
*  Second, the typical arithmetic operators, such as `+`, `-,` `*` and `/` have a different meaning for *regular Python lists* and *numpy arrays*.

Some lines of code have been provided for you. 

Try these out and select the one that would match this:

```
np.array([True, 1, 2]) + np.array([3, 4, False])
```

The `numpy` package is already imported as `np`.

In [38]:
np.array([True, 1, 2]) + np.array([3, 4, False])

array([4, 5, 2])

#### Instructions 4.3

**Possible answers**

$\bigotimes$  `np.array([True, 1, 2, 3, 4, False])`

$\large\checkmark$  `np.array([4, 3, 0]) + np.array([0, 2, 2])`

$\bigotimes$  `np.array([1, 1, 2]) + np.array([3, 4, -1])`

$\bigotimes$  `np.array([0, 1, 2, 3, 4, 5])`

---
## Exercise 4.4  Subsetting NumPy Arrays
---

Subsetting (using the *square bracket notation on lists or arrays*) works exactly the same with both lists and arrays.

This exercise already has two lists, `height_in` and `weight_lb`, loaded in the background for you. 

These contain the *height* and *weight* of the PBA players as *regular lists*. 

It also has two numpy array lists, `np_weight_lb` and `np_height_in` prepared for you.


In [71]:
## weight in pounds of PBA players since 1976
weight_lb = list([155, 165, 180, 175, 138, 194, 200, 200, 185, 170, 215, 180, 174,
       180, 160, 180, 175, 135, 210, 190, 205, 170, 188, 210, 146, 210,
       168, 185, 220, 210, 185, 215, 200, 200, 140, 170, 160, 180, 200,
       160, 195, 165, 177, 215, 160, 235, 180, 180, 200, 205, 149, 176,
       170, 175, 180, 200, 195, 220, 150, 165, 187, 170, 182, 180, 230,
       160, 225, 190, 205, 210, 180, 175, 180, 165, 212, 185, 210, 190,
       185, 172, 230, 220, 180, 180, 220, 221, 218, 210, 230, 180, 180,
       190, 258, 148, 180, 174, 160, 165, 170, 175, 200, 155, 190, 215,
       220, 170, 180, 200, 215, 165, 215, 224, 210, 170, 220, 185, 185,
       180, 180, 150, 195, 170, 185, 165, 175, 175, 143, 200, 165, 180,
       170, 182, 190, 180, 180, 260, 250, 185, 187, 175, 212, 200, 235,
       160, 179, 195, 205, 185, 200, 260, 190, 185, 200, 169, 185, 155,
       175, 170, 184, 212, 180, 170, 220, 185, 198, 205, 220, 186, 160,
       180, 165, 170, 185, 180, 176, 192, 220, 185, 175, 190, 170, 180,
       230, 222, 160, 200, 195, 190, 170, 170, 185, 210, 220, 200, 102,
       218, 237, 222, 206, 168, 170, 175, 135, 210, 205, 195, 178, 220,
       190, 190, 180, 215, 179, 180, 185, 195, 220, 227, 217, 200, 232,
       190, 210, 210, 200, 196, 200, 205, 245, 228, 200, 175, 300, 240,
       160, 160, 217, 250, 170, 180, 172, 200, 200, 180, 192, 185, 180,
       150, 184, 195, 230, 210, 200, 165, 185, 180, 155, 195, 208, 160,
       200, 180, 225, 187, 180, 210, 185, 170, 185, 163, 150, 150, 185,
       218, 179, 200, 185, 200, 250, 200, 210, 145, 205, 175, 160, 180,
       190, 185, 170, 210, 185, 165, 175, 210, 210, 185, 190, 205, 231,
       300, 142, 165, 220, 185, 170, 200, 215, 190, 180, 188, 185, 210,
       195, 180, 145, 151, 190, 210, 180, 168, 194, 180, 175, 150, 228,
       200, 165, 175, 170, 215, 200, 192, 200, 215, 185, 185, 184, 202,
       180, 215, 180, 155, 178, 165, 185, 190, 230, 168, 231, 245, 195,
       190, 188, 180, 195, 196, 150, 187, 155, 160, 240, 210, 180, 215,
       160, 170, 235, 145, 198, 220, 200, 151, 235, 170, 154, 180, 185,
       195, 190, 215, 205, 175, 199, 210, 218, 190, 200, 190, 200, 184,
       180, 185, 240, 179, 159, 180, 190, 198, 183, 210, 220, 252, 200,
       210, 201, 178, 170, 140, 189, 220, 230, 235, 205, 209, 170, 160,
       155, 192, 190, 176, 165, 195, 185, 180, 198, 180, 150, 150, 210,
       178, 190, 180, 180, 165, 225, 269, 193, 210, 185, 190, 210, 170,
       210, 145, 170, 180, 225, 180, 200, 170, 200, 148, 205, 175, 220,
       175, 170, 175, 215, 210, 160, 150, 185, 179, 220, 180, 165, 195,
       180, 175, 180, 180, 200, 210, 180, 160, 185, 190, 180, 165, 168,
       182, 175, 200, 185, 210, 215, 220, 179, 196, 205, 185, 200, 175,
       197, 300, 190, 237, 215, 180, 185, 170, 200, 200, 165, 193, 215,
       185, 244, 155, 245, 180, 143, 210, 170, 180])

In [72]:
print(type(weight_lb))
print(len(weight_lb))

<class 'list'>
516


#### Instructions 4.4

*  Subset `np_weight_lb` by printing out the element at index `50`.
*  Print out a sub-array of `np_height_in` that contains the elements at index `100` up to and including index `110`.

In [73]:
## code.py 4.4

import numpy as np

np_weight_lb = np.array(weight_lb)
np_height_in = np.array(height_in)

# Print out the weight at index 50
print(np_weight_lb[50])

# Print out sub-array of np_height: index 100 up to and including index 110
print(np_height_in[100:111])



149
[69 68 73 78 75 67 69 76 77 68 79]


---
## Exercise 4.5:  Your First 2D NumPy Array
---

Before working on the actual PBA data, let's try to create a 2D `numpy` array from a small list of lists.

In this exercise, `basketball` is a list of lists. 

The main list contains 4 elements. 

Each of these elements is a list containing the *height* and the *weight* of 4 baseball players, in this order. 

`baseball` is already coded for you in the script.

#### Instructions 4.5

*  Use `np.array()` to create a 2D `numpy` array from `basketball`. Name it `np_basketball`.

*  Print out the type of `np_basketball`.

*  Print out the `shape` attribute of `np_basketball`. Use `np_basketball.shape`.



In [79]:
## code.py 4.5

# Create basketball, a list of lists
basketball = [[180, 78.4],
            [215, 102.7],
            [210, 98.5],
            [188, 75.2]]

# Import numpy
import numpy as np

# Create a 2D numpy array from basketball: np_basketball
np_basketball = np.array(basketball)

# Print out the type of np_basketball
print(type(np_basketball))

# Print out the shape of np_basketball
print(np_basketball.shape)

<class 'numpy.ndarray'>
(4, 2)


---
## Exercise 4.6: Basketball data in 2D form
---

You realize that it makes more sense to restructure all this information in a 2D `numpy` array.

You have a Python list of lists. 

In this list of lists, each sublist represents the *height* and *weight* of a single basketball player. 

The name of this list is `player` and it will be loaded below.

Store the data as a 2D array to unlock `numpy`'s extra functionality.

In [160]:
# with open('datasets/player.pkl', 'wb') as file:
#    pickle.dump(player, file)

with open('datasets/player.pkl', 'rb') as file:
    player = pickle.load(file)

In [161]:
print(type(player))
print(len(player))
print(player[:3])

<class 'list'>
516
[[70, 155], [71, 165], [73, 180]]


#### Instructions 4.6

*  Use `np.array()` to create a 2D `numpy` array from `basketball`. Name it `np_basketball`.

*  Print out the *shape* attribute of `np_basketball`.

In [94]:
## code.py 4.6

import numpy as np

# Create a 2D numpy array from basketball: np_basketball
np_basketball = np.array(basketball)

# Print out the shape of np_basketball
print(np_basketball.shape)


(516, 2)


---
## Exercise 4.7:  Subsetting 2D NumPy Arrays
---

If your 2D `numpy` array has a regular structure, i.e. each *row* and *column* has a fixed number of values, complicated ways of subsetting become very easy. 

Have a look at the code below where the elements `"a"` and `"c"` are extracted from a list of lists.

```
# numpy
import numpy as np
np_x = np.array(x)
np_x[:, 0]
```

*  The indexes before the comma refer to the *rows*, while those after the comma refer to the *columns*. 

*  The `:` is for slicing; in this example, it tells Python to include all rows.


###  Instructions 4.7

*  Print out the 50th row of `np_basketball`.

*  Make a new variable, `np_weight_lb`, containing the entire second column of `np_basketball`.

*  Select the *height* (first column) of the 124th baseball player in `np_basketball` and print it out.


In [155]:
## code.py 4.7

import numpy as np

np_basketball = np.array(basketball)

# Print out the 50th row of np_basketball
print(np_basketball[49,0:])

# Select the entire second column of np_basketball: np_weight_lb
np_weight_lb = np_basketball[:,1]

# Print out height of 124th player
print(np_basketball[123,0])

[ 75 205  15]
70


---
## Exercise 4.7: 2D Arithmetic  
---

2D `numpy` arrays can perform calculations element by element, like `numpy` arrays.

`np_basketball` is coded for you; it's again a 2D `numpy` array with 3 columns representing *height* (in inches), *weight* (in pounds) and *years_played* (number of years played). 

`basketball` is available as a *regular list of lists* and `updated` is available as 2D `numpy` array.



In [152]:
import pickle

with open('datasets/basketball.pkl', 'rb') as file0:
    basketball = pickle.load(file0)

with open('datasets/updated.pkl', 'rb') as file1:
    updated = pickle.load(file1)


In [153]:
print(type(basketball))
print(len(basketball))
print(basketball[:3])

<class 'list'>
516
[[70, 155, 1], [71, 165, 12], [73, 180, 6]]


In [154]:
print(type(updated))
print(updated.shape)
print(updated[:3])

<class 'numpy.ndarray'>
(516, 3)
[[  0.430988    24.40705198   1.        ]
 [  0.91349429 -31.39857149   1.        ]
 [  0.5890831   19.97431077   1.        ]]


## Instructions 4.7

*  You managed to get hold of the changes in `height`, `weight` and `years_played` of all basketball players. It is available as a 2D numpy array, `updated`. Add `np_baseball` and `updated` and print out the result.

*  You want to convert the units of height and weight to metric (*meters* and *kilograms*, respectively). As a first step, create a `numpy` array with three values: `0.0254`, `0.453592` and `1`. Name this array `conversion`.

*  Multiply `np_baseball` with `conversion` and print out the result.


In [162]:
## code.py 4.7

# basketball is available as a regular list of lists
# updated is available as 2D numpy array

# Import numpy package
import numpy as np

# Create np_baseball (3 cols)
np_basketball = np.array(basketball) 

# Print out addition of np_basketball and updated
print(np_basketball + updated)

# Create numpy array: conversion
conversion = np.array([0.0254, 0.453592, 1])

# Print out product of np_basketball and conversion
print(np_basketball * conversion)


[[ 70.430988   179.40705198   2.        ]
 [ 71.91349429 133.60142851  13.        ]
 [ 73.5890831  199.97431077   7.        ]
 ...
 [ 78.59166301 180.78487452   9.        ]
 [ 71.53708962 161.17811067  10.        ]
 [ 72.28368578 213.45440371   2.        ]]
[[ 1.778   70.30676  1.     ]
 [ 1.8034  74.84268 12.     ]
 [ 1.8542  81.64656  6.     ]
 ...
 [ 1.9812  95.25432  8.     ]
 [ 1.8034  77.11064  9.     ]
 [ 1.8034  81.64656  1.     ]]


---
## Exercise 4.8: Average versus median
---

You now know how to use `numpy` functions to get a better feeling for your data.

The `basketball` data is available as a 2D numpy array with 3 columns (`height`, `weight`, `years_played`) and 516 rows. 

The name of this `numpy` array is `np_basketball`. 

After restructuring the data, however, you notice that some height values are abnormally high. 

Follow the instructions and discover which summary statistic is best suited if you're dealing with so-called *outliers*.` np_basketball` is available.


In [163]:
## code.py 4.8

# Import numpy
import numpy as np

# Create np_height_in from np_basketball
np_height_in = np_basketball[:,0]

# Print out the mean of np_height_in
print(np.mean(np_height_in))

# Print out the median of np_height_in
print(np.median(np_height_in))

74.44961240310077
75.0


---
## Exercise 4.9:  Explore the basketball data
---

The Python script in the editor already includes code to print out informative messages with the different summary statistics and `numpy` is already loaded as `np`. 

Can you finish the job? `np_basketball` is available.

## Instructions 4.9

*  The code to print out the *mean height* is already included. Complete the code for the *median height*. Replace `None` with the correct code.

*  Use `np.std()` on the first column of `np_basketball` to calculate `stddev`. Replace `None` with the correct code.

*  Do big players tend to be heavier? Use `np.corrcoef()` to store the correlation between the first and second column of `np_basketball` in `corr`. Replace `None` with the correct code.



In [164]:
## code.py 4.9

# np_basketball is available

# Import numpy
import numpy as np

# Print mean height (first column)
avg = np.mean(np_basketball[:,0])
print("Average: " + str(avg))

# Print median height. Replace 'None'
med = np.median(np_basketball[:,0])
print("Median: " + str(med))

# Print out the standard deviation on height. Replace 'None'
stddev = np.std(np_basketball[:,0])
print("Standard Deviation: " + str(stddev))

# Print out correlation between first and second column. Replace 'None'
corr = np.corrcoef(np_basketball[:,0],np_basketball[:,1])
print("Correlation: " + str(corr))






Average: 74.44961240310077
Median: 75.0
Standard Deviation: 3.324121939922492
Correlation: [[1.         0.76554252]
 [0.76554252 1.        ]]
